### ISSUE TITLE
Method naming scheme

### ISSUE THREAD

**Issue by mpmc** [jarvisteach/appJar#235]
When writing classes, I try to do method names as simple as possible.. e.g.

```
def do():
   return value
def do_set(val):
    value = val
```
or when I group methods..

```
def main_foo():
    pass

def main_bar():
    pass
``` 

I noticed that appJar is using CamelCase, with add/set etc at the start of method names, is there a reason for the non pep-8 usage? **I'm not saying this way is wrong**, I just find it harder to read.

For example addLabelEntry/addLabelNumericEntry could be replaced with entry() and numeric_entry(), To get values you could just simply use entry_get() which for me is a lot easier to read and avoids camels.

**All this of course is a matter of personal taste.**

**Comment by jarvisteach** [comment]
I've had this discussion with myself a few times.

I grew up on camel case - and have always used it - I've never worked for a company that insisted on underscores in their style guides. It's also what I teach in the classroom, and recommend my kids use. It's also more close to what most of the other languages in my classroom look like (Java, VB, C#, SmallBasic,, C++) - so makes for easier transitions...

When I started teaching python a couple of years ago, I had no idea there was such a thing as a PEP, and I had no idea python recommended using underscores. I hadn't come across them in any of the basic python commands or libraries, in the turtle package (apart from begin_fill and end_fill - which always looked so out of place), or in the basic tkinter commands. So I stuck with what I knew.

As I've never used underscores, I find them hard to read and annoying, and hard to teach - it's an unfamiliar key for young kids, and an extra key press when they're typing...

So, I've convinced myself camel case is better - obviously through my rose tinted glasses...

Now that I'm enjoying the micro:bit, I don't really find the underscores an improvement - they make it hard to fit code on the screen as lines get so long... But, it would be nice to have some consistency across the things I do with Python.... And, if appJar becomes useful to people, I suppose it would be nice to comply with the python style guides...

As to the function names - again they're like that from a teaching point of view - I wanted consistency along with self-explanatory code. A lot of them have been driven straight from the kids' expectations; once they get the basics, they start writing their GUIs and just expect commands to be there, following the same syntax - so I add them in...

The same is true of parameters. The kids struggle at first with parameters, with named parameters being another step up - so I have lots of set functions, instead of lots of named parameters.

Ultimately - the library has been designed for 11 year old kids. I don't want to lose the ease of use for them, but at the same time I would like it to become more friendly for developers...

I could have duplicate function names, to offer both styles, but that seems like overkill, especially with the size of the file already (again - it's all in one file to make it as easy as possible to get going).

So I'm unsure what to do...

It would be a massive task to get it changed, and with me back in school next week - there's not going to be much time for changes!

**Comment by mpmc** [comment]
> I grew up on camel case - and have always used it
As did I, I got used to seeing it & even using it when I had to, but to me it felt wrong.  I started with VB, but soon moved on to PHP, which I've been writing since I was 12!

And I'm really glad to see programming languages being taught now, when I was at (high) school, we were just taught about word processing, Microsoft Word, Excel, Powerpoint etc. I cringe every time I think of, or see "mailmerge".  This is at a time when we had mixed Windows 95, 98, Windows 2000, XP AND 3.1 with networking in different rooms.

IIRC I was once asked to "Design a website in powerpoint", which to me made absolutely no sense! It was extremely boring & a real pain for me (you'll find out why as you read on). I did it using PHP instead. Of course I failed, the teacher had no idea what PHP was, or simply refused to acknowledge it existed.

> As I've never used underscores, I find them hard to read and annoying, and hard to teach - it's an unfamiliar key for young kids, and an extra key press when they're typing...

I can see why you prefer camel case, but for me capitals are a pain in the backside, mostly because I only use one hand and have to do a lot of moving and stretching to one side of the keyboard to reach modifiers & the button at the same time. 

> they make it hard to fit code on the screen as lines get so long... 

Yes, this can be true in many cases, I try and keep my code =< 80 lines for this reason. But with Python long lines aren't so much of a problem, for example.

```
        self.setSticky('nesw')
        self.addLabel('FOOOOOOOOBAAAAARRRRRRRRRR', 
                      self.lang('WOOOOOO_HOOOO'), 0, 0)
```
But this is again down to personal preference :).

> So I'm unsure what to do...
Continue as is! I and I'm sure many others will get used to it!

---
This is what my current code looks like using appJar. https://gist.github.com/mpmc/24f055b88be547c62a6d35fe51e63cf5

**Comment by keithterrill** [comment]
My two cents regarding naming scheme.

1. Sir, you are a teacher, instructor. I believe that the simplicity of your code and the project is reflected in this fact. As you go forward and ponder your statement, "So I'm unsure what to do..." I recommend you keep KISS (Keep It Simple Silly) in mind.

2. The PEP-8 (https://www.python.org/dev/peps/pep-0008/) reminds us, "do not break backwards compatibility just to comply with this PEP!" It also tells us, "A style guide is about consistency. Consistency with this style guide is important. Consistency within a project is more important. Consistency within one module or function is the most important."

3. mpmc may find that CamelCase is harder to read, but there are many of us who do not find that to be the case. I too am old school. I began programming in 1976.

4. Yes, addLabelNumericEntry could be replaced with "entry" or "numeric_entry", but at a loss of information. addLabelNumericEntry tells me 4 things: 1) we are adding the widget not reopening, 2) it contains a label in the grid cell, 3) it is a numeric not a string entry, and 4) it is an Entry field. It appears that the only way to change this while keeping the information in the name would be to use add_label_numeric_entry, which is cumbersome and takes up more real estate.

5. Your current style has produced a very readable and self documenting project.

6. It is my humble opinion that you do not need to change the style, nor should you. As you point out, "It would be a massive task to get it changed..."

7. By the way: very good work, job done well.

**Comment by jarvisteach** [comment]
Thank you both for the support - I have some longer term aims, in terms of refactoring certain parts (widget management being the big one) - now looking more like V2.0 stage - and I will probably look into this sort of stuff then...

**Comment by jarvisteach** [comment]
So, I've been giving this some more thought.

I think there are now too many functions for doing very similar things. Using entry (the busiest of the widgets) as an example:
```
app.addEntry()
app.addLabelEntry()
app.addValidationEntry()
app.addNumericEntry()
app.addDirectoryEntry()
app.addSecretEntry()
app.addFileEntry()
app.addAutoEntry()
```

Combining this with all the SET & GET functions I think is starting to make appJar quite difficult to get to grips with.

What I'm thinking is providing a single function per widget:
```
app.entry(title, value=None)
```

This will then serve all three functions - ADD/GET/SET.  
It can use a bit of logic to determine what the user wants:
```
IF title exists:
    IF value is None: GET
    ELSE: SET
ELSE:
    ADD
```

This will provide a nice consistent approach, and achieves almost as much as we have now. The only thing lacking is the warning of duplicate widgets. If you are trying to ADD a widget, but use a duplicate name, you will mistakenly SET the original widget. I think this is an acceptable trade-off.

It can then wrap up all the different options using optional `**kwargs`:
```
secret=False # make it secret
type in ["validation", "directory", "numeric", "file", "auto"] # choose the type
default=None # set a default
label=False # show a label
limit=None # set a length limit
focus=False # give it the focus
case in ["upper", "lower"] # make all text upper/lower case
```

I think this is very achievable, and for now, can be added as a wrapper around most of the other functions, and run in parallel with what we already have.

---

Stage two is to look at all the set functions that get magically created. Again, I think these are getting unmanageable. I think a `configEntry()` function could be provided using `**kwargs` to set all the properties. I think this will be done under a separate change request.

---

Not sure what the longterm plan should be - will definitely implement these and run them in parallel, with documentation. Perhaps medium term, if they prove popular, will look to deprecate the old functions and remove previous documentation. Will hold out on that for now though...

**Comment by mpmc** [comment]
This is a nice idea, and the fact it's going to be backwards compatible is even better, as I've gotten used to the method names already in place!

> This will provide a nice consistent approach, and achieves almost as much as we have now. The only thing lacking is the warning of duplicate widgets. If you are trying to ADD a widget, but use a duplicate name, you will mistakenly SET the original widget. I think this is an acceptable trade-off.

I think this is what some people expect anyway.

> It can then wrap up all the different options using optional **kwargs:

This is pretty much what I do with kwargs.

```python
def foo(param, **kwargs):
    print(param)
    optional = kwargs.get('weird') or False
    if optional is True:
        return print('I\'m weird!')
    print('I\'m not weird!')
```

**Comment by jarvisteach** [comment]
All basic widgets now implemented - couple of issues to resolve.

Most of the widget specific set functions now implemented as parameters to the ADD function:
`app.entry("name", default="Enter username", focus=True, limit=12)`

Don't support labelled widgets yet, would like a boolean `label` flag to show a label. But want to refactor labelelled widget creation, rather than have conditional constructor calls.

Documentation started here (needs a lot of work): https://github.com/jarvisteach/appJar/blob/next_release/docs/mkdocs/docs/simpleAppJar.md

**Comment by mpmc** [comment]
@jarvisteach 

I'm currently rewriting my hs602/controller class (for a 5th time :frowning: ) after discovering the [@property decorator](https://docs.python.org/2/library/functions.html#property), and wondered if you used it in appJar, took a look & saw that you don't so decided to mention it here. I'm finding it extremely useful, hope you do too :+1:

**Comment by jarvisteach** [comment]
Yeah - I've had a think about this. And originally wanted to use it, but couldn't see how it would fit...

The point of the property decorator is to allow you to access getter/setter functions on attributes by simply referencing their name.

But for that to work, setting should require a single value and getting should require None.

So, for example:
```
gui.label = "Hello World"
print ( gui.label )
```

That model doesn't really work in appJar, as both of the above functions would also require the widget's name as a parameter.

So, for setters, we could solve it by passing a tuple of values and unpacking them:
```
# original style
app.addLabel("a", "Hello")
app.setLabel("a", "new hello")
print ( app.getLabel("a" )

# new style being introduced
app.label("b", "Hello")
app.label("a", "new hello")
print ( app.label("a" )

# as a property setter
app.label = ("c", "hello")
app.label = ("c", "hello")
print ( app.label("c") )
# or
app.label = "c", "hello"
app.label = "c", "hello"
print ( app.label("c") )
```

I think each one is less *self-documenting* than the previous, and less intuitive.

I could have skipped the new style being introduced and wrapped the `getLabel` & `addLabel` up as properties, but I really don't like the syntax necessary to support the extra parameter...

**Comment by mpmc** [comment]
I see, I honestly didn't see any use for it either, but I stuck with it..

This is what my code looks like now using properties (still WIP) - I'm just glad I could reuse most of the existing code & tweak things as needed.

```Python
import socket


class Hs602(object):
    """HS602 controller."""
    def __init__(self, **kwargs):
        # Defaults.
        self._addr = '<broadcast>'
        self._udp = 8086
        self._tcp = 8087
        self._encoding = 'utf-8'
        self._ping = 'HS602'
        self._pong = 'YES'
        self._cmd_len = 15
        self._timeout = 10
        self.__socket = None

        # Allow kwargs override.
        for key, value in kwargs.items():
            if hasattr(self, "_{}".format(key)):
                setattr(self, key, value)

    def addr_get(self):
        """Device address."""
        return str(self._addr)

    def addr_set(self, value):
        """Set device address - will kill existing sockets.

        :param value: Device (or broadcast) address to set.
        """
        self._addr = str(value)
        # Close any existing sockets if addr is set.
        self.close()

    addr = property(addr_get, addr_set)

    def udp_get(self):
        """UDP broadcast port."""
        return int(self._udp)

    def udp_set(self, value):
        """Set UDP broadcast port - will kill existing sockets.

        :param value: UDP to broadcast on, uses default if not 1-65535.
        """
        if value not in range(1, 65536):
            value = 8086
        self._udp = int(value)
        # Close any existing sockets if udp is set.
        self.close()

    udp = property(udp_get, udp_set)

    def tcp_get(self):
        """TCP command port."""
        return int(self._tcp)

    def tcp_set(self, value):
        """Set TCP command port - will kill existing sockets.

        :param value: TCP command port, uses default if not 1-65535.
        """
        if value not in range(1, 65536):
            value = 8087
        self._udp = int(value)
        # Close any existing sockets if tcp is set.
        self.close()

    tcp = property(tcp_get, tcp_set)

    def encoding_get(self):
        """Message encoding."""
        return str(self._encoding)

    def encoding_set(self, value):
        """Set message encoding.

        :param value: Encoding value.
        """
        self._encoding = str(value)

    encoding = property(encoding_get, encoding_set)

    def ping_get(self):
        """Ping trigger message."""
        return str(self._ping).encode(self.encoding)

    def ping_set(self, value):
        """Set ping trigger message.

        :param value: Ping message.
        """
        self._ping = str(value)

    ping = property(ping_get, ping_set)

    def pong_get(self):
        """Expected pong reply."""
        return str(self._pong).encode(self.encoding)

    def pong_set(self, value):
        """Set pong reply.

        :param value: Expected pong reply.
        """
        self._pong = str(value)

    pong = property(pong_get, pong_set)

    def cmd_len_get(self):
        """Command length."""
        return int(self._cmd_len)

    def cmd_len_set(self, value):
        """Set expected command length.

        :param value: Command length.
        """
        self._cmd_len = int(value)

    cmd_len = property(cmd_len_get, cmd_len_set)

    def timeout_get(self):
        """Socket timeout."""
        return int(self._timeout)

    def timeout_set(self, value):
        """Set socket timeout.

        :param value: Timeout value 1-500.
        """
        if value not in range(1, 501):
            value = 10
        self._udp = int(value)

    timeout = property(timeout_get, timeout_set)

    def devices_get(self):
        """Return a list of discovered devices.
           This will send a ping (to self.addr:self.udp) & wait for
           replies until timeout is reached - it will block!

        """
        replies = self._udp_msg(self.ping)
        if replies:
            return [rep[0] for rep in replies if rep[2] == self.pong]

    devices = property(devices_get)

    def properties_get(self):
        """Return a dictionary containing command properties."""
        values = {
            'rtmp': {
                'url': 16,
                'key': 17,
                'username': 20,
                'password': 21,
            },
            'colour': {
                'brightness': 0,
                'contrast': 1,
                'hue': 2,
                'saturation': 3,
            },
            'source': {
                2: 'ypbpr',
                3: 'hdmi',
            },
            'resolution': {
                0: '1920x1080 60Hz',
                1: '1280x720 60Hz',
                2: '720x480 60Hz',
                3: '720x480 60Hz',
                4: '720x480 60Hz',
                5: '1920x1080 50Hz',
                6: '1280x720 50Hz',
                7: '720x576 50Hz',
                8: '720x576 50Hz',
                9: '720x576 50Hz',
                10: '1920x1080 60Hz',
                11: '1280x720 60Hz',
                12: '720x480 60Hz',
                13: '720x480 60Hz',
                14: '1920x1080 50Hz',
                15: '1280x720 50Hz',
                16: '720x576 50Hz',
                17: '720x576 50Hz',
                18: '720x480 60Hz',
                19: '720x576 50Hz',
                20: '1920x1080 25Hz',
                21: '1920x1080 30Hz',
                22: '0x0 60Hz',
                23: '640x480 60Hz',
                24: '1920x1080 30Hz',
                25: '1920x1080 25Hz',
                26: '1920x1080 50Hz',
                27: '1920x1080 60Hz',
                28: '1920x1080 24Hz',
                29: '1920x1080 60Hz',
                30: '1920x1080 50Hz',
                31: '1920x1080 24Hz',
                32: '800x600 60Hz',
                33: '1024x768 60Hz',
                34: '1152x864 60Hz',
                35: '1280x768 60Hz',
                36: '1280x800 60Hz',
                37: '1280x960 60Hz',
                38: '1280x1024 60Hz',
                39: '1360x768 60Hz',
                40: '1440x900 60Hz',
                41: '1600x900 60Hz',
                42: '1680x1050 60Hz',
            },
        }
        # Color/Colour workaround.
        values['color'] = values['colour']
        return values

    properties = property(properties_get)

    def _pad(self, data, pad_len=None):
        """Return data appended with zero bytes to the size of pad_len.

        :param data: data to pad, can be anything that can be
        converted to bytes.
        :param pad_len: Size required, don't set to use self.cmd_len.
        """
        pad_len = pad_len or self.cmd_len
        data = bytes(data)
        data = data.ljust(pad_len, b'\0')
        return data

    def _echo(self, first, second):
        """Check if two variables are equal.

        :param first: First object.
        :param second: Second object.
        """
        return first == second

    def _udp_msg(self, msg, reply=True):
        """Send UDP message.

        :param msg: Message (in bytes) to send.
        :param reply: Do we want to wait for a reply?

        Return a list of replies, an empty list is returned if
        nothing is received.
        """
        msg = bytes(msg)
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            s.settimeout(self.timeout)
            s.bind(('', self.udp))

            replies = []
            while True:
                # Send message.
                if msg:
                    sent = s.sendto(msg, (self.addr, self.udp))
                    if not sent > 0:
                        break
                    msg = msg[sent:]
                    continue

                # Receive message?
                if not reply:
                    return True
                try:
                    data, [addr, port] = s.recvfrom(2048)
                    replies += [[addr, port, data]]
                except socket.timeout:
                    break
            return replies

    def _send(self, msg, reply=True):
        """Send command to device - If no address is set self.discover
        will be used.

        :param msg: Message (in bytes) to send.
        :param reply: Do we want to wait for a reply?

        Return the reply. True if
        message was sent without error (when no reply needed).
        """
        msg = bytes(msg)
        # Do we need to discover a device first?
        if not self.addr:
            self.addr = self.discover[0]
        # If there is no connection, so tell the device to
        # get ready to accept one.
        #
        # This is just the the "C" char and the IPv4 address of the
        # box in reverse.
        if not self.__socket:
            ip = reversed(self.addr.split('.'))
            cmd = [67] + [int(octal) for octal in ip]
            self.udp_msg(cmd, False)
            addr = (self.addr, self.tcp)
            self.__socket = socket.create_connection(addr, self.timeout)

        # Send it & get replies (if needed).
        self.__socket.sendall(msg)
        if not reply:
            return True
        data = bytes()
        while True:
            buf = self.__socket.recv(self.cmd_len)
            if not buf:
                self.close()
                break
            data += buf
            if len(data) == self.cmd_len:
                break
        return data

    def close(self):
        """Kill the connection."""
        try:
            # If this fails we can ignore it.
            self.__socket.shutdown(socket.SHUT_RDWR)
            self.__socket.close()
        except (OSError, AttributeError) as exc:
            pass
        self.__socket = None

    def socket_get(self):
        """Return True if the socket _hasn't_ need shutdown."""
        try:
            return True if self.socket else None
        except AttributeError as exc:
            pass

    socket_active = property(socket_get)
```

**Comment by mpmc** [comment]
Trying the new renamed widgets, getting an error with list.

```
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.6/tkinter/__init__.py", line 1702, in __call__
    return self.func(*args)
  File "/usr/lib/python3.6/tkinter/__init__.py", line 748, in callit
    func(*args)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 1903, in __processEventQueue
    func(*args, **kwargs)
  File "gui2.py", line 278, in result
    self.list('addr', devices, row=1, column=0)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 6208, in list
    return self.listBox(title, value, *args, **kwargs)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 6227, in listBox
    if change is not None: self.setListBoxChangeFunction(title, change)
  File "<string>", line 1, in setListBoxChangeFunction
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 2529, in configureWidget
    self.__bindEvent(kind, name, item, value, option, key)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 2695, in __bindEvent
    cmd = self.MAKE_FUNC(function, name, True)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 3069, in MAKE_FUNC
    raise Exception("Invalid function: " + str(funcName))
Exception: Invalid function: False
```

I'm also not sure on calling it list, it could be confused with pythons own list()?

Example being
```python
list(app.list(foo, bar))
```

**Comment by jarvisteach** [comment]
Fixed the error - was defaulting to False instead of None.

I agree on the naming - there are a couple in there I'm unsure of. I'm trying to keep each widget to a single word, any suggestions? There are also some param names that will probably change (type)...

**Comment by mpmc** [comment]
Hmm, tough choice on the name, I'd just call it what it is.. a listbox. :grin: 

Just throwing some ideas out - feel free to ignore, how about having each widget as a separate class that sub-classes?
```python
class AppJar(Tk):

    class Widget(object):
        ...

    class Listbox(Widget):
        ...
        
app = AppJar(...)
app.Listbox(...)
```

**Comment by mpmc** [comment]
- [x] addLabel - label
- [x] addEntry - entry
- [x] addButton - button
- [x] addCheckBox - check - should we have checkbox, checkBox exists?
- [x] addRadioButton - radio - ditto
- [x] addOptionBox - option - ditto
- [x] addSpinBox - spin - ditto
- [x] addDatePicker - date - ditto
- [x] addListBox - list - this might be confused with Pythons own list?
- [x] addProperties - properties - this may get confused with pythons own properties?
- [x] addScale - scale
- [x] addMessage - message
- [x] add{Horizontal,Vertical}Separator - separator
- [x] addTextArea - text
- [x] addMeter - meter
- [x] addLink - link
- [x] addGrip - grip

---

Some ideas:-

- [ ] Allow inside padding to be set when calling widget? E.g. app.label(padding=[x, x]) or app.label(pad_y=x, pad_x=x) Same with stretch & sticky?
- [ ] Some methods could use properties, so instead of app.setTitle() use ```app.title = "My app title"```. I'm not 100% sure on this one, but throwing it out there anyway.

**Comment by jarvisteach** [comment]
Hi @mpmc, I have included functions for `radioButton`, `optionBox`, etc - they're just not documented. if people intuitively try to use them, they can, but I'm keen to ge single word function names...

I'm not too worried about `properties` - I think that one's OK.

But, yes `list` I'm very unhappy with. There is a `app.listBox()` which works, but that does need resolving...

I've spent the last 3 weeks teaching my year 12s to build GUIs in C#.Net, so I'm thinking it would be nice to have widget names match up to those as well...

Not sure what I should change on the message boxes, they are slightly different to widgets, and don't have the add/get/set options?

One thing I'm planning is to add in all the config options! So, you can pass as many parameters as you want into each function, any left at the end will be farmed off to a config function, that will try to apply them to the widget using the widget's `.config()` function. I want this to work when both adding & setting, so it's going to need a bit of work, but I'll add a quick demo for labels. So that will tick your first  box.

I think the second suggestion should be a new feature request. I like the idea in general, but it will need a bit of research to build a list, and then a bit of work to sort it out.

**Comment by jarvisteach** [comment]
I've just upgraded `.label()` so you should now be able to set `bg="red"`, etc...

**Comment by mpmc** [comment]
> Hi @mpmc, I have included functions for radioButton, optionBox, etc - they're just not documented. if people intuitively try to use them, they can, but I'm keen to ge single word function names...

Ah, I see :)

> But, yes list I'm very unhappy with. There is a app.listBox() which works, but that does need resolving...

It's a list of options so.

* choices
* select / selector

I like choices best.

> Not sure what I should change on the message boxes, they are slightly different to widgets, and don't have the add/get/set options?

Sorry, what I meant is have a single word method for those too, popup looks good to me :)

So instead of calling ```yesNoBox()``` use  ```popup('foo', 'bar', type='yesno')``` with the default type being info? ```textBox()``` & ```numberBox()``` could be  ```prompt('foo', 'bar', type=[int, 'number'])```, default being string?

**Comment by jarvisteach** [comment]
Ahhh, yes! I like the popUps idea, I'll add that in!

**Comment by mpmc** [comment]
updated & I'm getting ```2017-12-15 00:30:56,462 appJar:ERROR [Line 157->7075/_configWidget]: Unexpected **kwarg for widget text1: {'sticky': 'nesw', 'stretch': 'both'}```

for..

```python
self.label('text1', text, sticky='nesw', stretch='both')
```
sticky/stretch not added yet?

**Comment by jarvisteach** [comment]
Yup, those two properties are applied at the container level - so they're not specific to a widget.

The container's stickiness can be overridden when the widget is positioned, so there's an option to configure that. The stretchiness, can't currently be provided as a parameter, so that'll require a bit more work...

**Comment by mpmc** [comment]
@jarvisteach Ah, that makes sense!

**Comment by jarvisteach** [comment]
@mpmc BTW - if you want to keep up to date with the features, the docs can be viewed here: https://github.com/jarvisteach/appJar/blob/next_release/docs/mkdocs/docs/simpleAppJar.md

**Comment by mpmc** [comment]
Getting 
```python
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.6/tkinter/__init__.py", line 1702, in __call__
    return self.func(*args)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 3164, in <lambda>
    return lambda: funcName(param)
  File "gui3.py", line 191, in manual
    self.entry('tcp', 1, 1, kind='numeric', limit=5)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 7514, in entry
    if value is not None: self.setEntry(title, value)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 7870, in setEntry
    if not ent.isNumeric or self.__validateNumericEntry("1", None, text, None, "1", None, None, None):
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 7824, in __validateNumericEntry
    if len(str(value_if_allowed)) == 1 and value_if_allowed in '.-':
TypeError: 'in <string>' requires string as left operand, not int
```
using 

```python
self.entry('tcp', 1, 1, kind='numeric', limit=5)
self.entry('timeout', 3, 1, kind='numeric', limit=5)
```

**Comment by mpmc** [comment]
option() doesn't accept focus like the entries? Not sure this should even be possible?

```python
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.6/tkinter/__init__.py", line 1702, in __call__
    return self.func(*args)
  File "/usr/lib/python3.6/tkinter/__init__.py", line 748, in callit
    func(*args)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 1943, in __processEventQueue
    func(*args, **kwargs)
  File "gui3.py", line 156, in result
    self.option('addr', devices, 1, 0, focus=True)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 4788, in option
    return self.optionBox(title, value, *args, **kwargs)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 4801, in optionBox
    else: opt = self.addOptionBox(title, value, *args, **kwargs)
TypeError: addOptionBox() got an unexpected keyword argument 'focus'
```

**Comment by mpmc** [comment]
I've just tried to pass align to a label but it's not accepted :(.

e.g.
```python
label('foo', 'bar', align='center')
```

**Comment by mpmc** [comment]
Latest code breaks :(

```python
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.6/tkinter/__init__.py", line 1702, in __call__
    return self.func(*args)
  File "/usr/lib/python3.6/tkinter/__init__.py", line 748, in callit
    func(*args)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 1945, in __processEventQueue
    func(*args, **kwargs)
  File "gui3.py", line 165, in callback
    self.button(_('\u26B2 Scan again'), self.discover, 2, 0, 1)
  File "/home/mark/.local/lib/python3.6/site-packages/appJar/appjar.py", line 6683, in button
    else: button = self._buttonMaker(title, value, "button", extra=None, *args, **kwargs)
TypeError: _buttonMaker() got multiple values for argument 'extra'
```

**Comment by jarvisteach** [comment]
@mpmc I don't think `align` is a valid argument, try setting `anchor="e"` - there's better error handling in the newer commits.

The second error is an interesting one - it has nothing to do with the `extra` parameter, and is in fact because I've named some parameters, but then *args has some unnamed parameters - I'll fix that...

**Comment by mpmc** [comment]
Ok thanks ![](https://assets-cdn.github.com/images/icons/emoji/unicode/1f44d.png)
I'll wait for you to release 0.90 as I have some new ideas for my GUI :smile:

**Comment by jarvisteach** [comment]
Tracking normalisation of simpleGUI:

- [ ] label
- [ ] entry
- [ ] button
- [ ] check
- [ ] radio
- [ ] option
- [ ] spin
- [ ] date
- [ ] listBox
- [ ] properties
- [ ] scale
- [ ] message
- [ ] separator
- [ ] text
- [ ] meter
- [ ] link
- [ ] grip
- [ ] various message boxes - YesNoBox, InfoBox etc.

**Comment by jarvisteach** [comment]
Must also ensure `callFunction` is propagated to setters.
